src/snowflake/cli/api/utils/graph.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+
+
+class Graph:
+    @dataclass
+    class Node:
+        key: str
+        status: str | None = None
+        dependencies: set[Graph.Node] = field(default_factory=set)
+
+        def __eq__(self, other):
+            return self.key == other.key
+
+        def __hash__(self):
+            return hash(self.key)
+
+    def __init__(self):
+        self._graph_nodes_map: dict[str, Graph.Node] = {}
+
+    def get(self, key: str):
+        if key in self._graph_nodes_map:
+            return self._graph_nodes_map.get(key)
+        raise KeyError(f"Node with key {key} not found")
+
+    def get_all_nodes(self):
+        return self._graph_nodes_map.values()
+
+    def add(self, node: Node):
+        if node.key in self._graph_nodes_map:
+            raise KeyError(f"Node key {node.key} already exists")
+        self._graph_nodes_map[node.key] = node
+
+    def add_dependency(self, key1: str, key2: str):
```

Comment:
nit: this is a directed graph, 1 and 2 are unclear, rename to include the direction of the edge (some kind of from/to naming)